### DISCUSSION TITLE
Documentation on custom actions -- and custom actions quiet execution in general

### DISCUSSION THREAD

**Discussion by bbmmcodegirl** [wixtoolset/issues#8195] [Questions]
The documentation on quiet execution of custom actions is here: https://wixtoolset.org/docs/tools/wixext/quietexec/

It is very confusing. This may be due to the actual implementation of the feature being confusing.

Just for example, it says about executing an immediate custom action quietly:

* As example: to create a property (with `SetProperty`) with the Id `WixQuietExecCmdLine` to be scheduled `Before` the relevant custom action (in the example, called `ConvolutedCoreCountCustomAction`). 
* There is nothing to tell how the property and the custom action belong together and are linked, except that the `SetProperty` is scheduled `Before` the `CustomAction`.
* The `DllEntry` of the `CustomAction` is supposed to have the value `"WixQuietExec"`. This doesn't make any sense: The documentation [here](https://wixtoolset.org/docs/v3/xsd/wix/customaction/) states "This attribute specifies the name of a function in a custom action to execute." In case my custom action is to execute a function in a `dll`, then I already need to use the `DllEntry` to specify which function, I can't stick another value in. How am I supposed to do that?

**Comment by robmen** [comment]
I'm confused. What are you trying to do?

**Comment by bbmmcodegirl** [comment]
Running a custom action that is a method in a dll quietly. 
Dll: MyDll.dll
Method: MyFunction in MyDll.dll
And how to run this with WixQuietAction?
Please state how I should express that in Wix (the xml of it).

**Reply by bevanweiss** [reply]
If you have your own CustomAction, then you would NOT use the WixQuietExec CustomAction, you would reference your own CustomAction.

WixQuietExec **IS** is a CustomAction (written by the WiX team, which executes a command line 'quietly').
This is why it needs the WixQuietExecCmdLine updated before it executes, WixQuietExec just runs the command line WixQuietExecCmdLine.

If you have authored your own custom action, then you should be using this:
https://wixtoolset.org/docs/schema/wxs/customaction/
and likely this
https://wixtoolset.org/docs/schema/wxs/binary/

And then your CustomAction will execute as quietly, OR as loudly, as YOU have authored it.

**Reply by bbmmcodegirl** [reply]
Can you please provide the link to the documentation , or a code example of the wix xml to use, for WixQuietExecCmdLine  and WixQuietExec  ?

**Reply by bevanweiss** [reply]
https://github.com/wixtoolset/wix3/blob/b40e9a32c24033e11b77baf2c91a704382f898ed/src/chm/documents/customactions/qtexec.html.md?plain=1#L25-L34

**Comment by bbmmcodegirl** [comment]
Hi @bevanweiss I am doing exactly what that documentation says I should, but my custom actions do not run. I am doing the following:
`<Property Id="MyCustomActionId" Value='MyCommandLine' />`
`<CustomAction Id="MyCustomActionId" BinaryKey="WixCA" DllEntry="WixQuietExec" Execute="deferred" Return="ignore" Impersonate="no"/>`
`<InstallExecuteSequence>`
`      <Custom Action="MyCustomActionId" After="InstallFiles"></Custom>`
`</InstallExecuteSequence>`

and I have tried the following as `MyCommandLine`:
`"dir"` (with the double quotes)
`"cmd /C Dir"` (with the double quotes)

And in either case the custom action fails with this in the installation logfile:
`WixQuietExec:  Entering WixQuietExec in C:\WINDOWS\Installer\MSI10E5.tmp, version 3.11.4516.0`
`WixQuietExec:  "cmd /C Dir"`
`WixQuietExec:  Error 0x80070002: Command failed to execute.`
`WixQuietExec:  Error 0x80070002: QuietExec Failed`
`WixQuietExec:  Error 0x80070002: Failed in ExecCommon method`

The documentation should have a real example for how you have to enter a simple script to execute quietly.

**Reply by bevanweiss** [reply]
This aligns with my expectations... what are you expecting?
![image](https://github.com/wixtoolset/issues/assets/2055446/42483db9-8802-4c38-ae84-fc8508180c58)

Usage might be similar to (you can search in github for other usages)
```xml
    <!-- Custom Actions to Install Chocolatey -->
    <SetProperty Id="InstallChocolatey"
      Before="InstallChocolatey"
      Sequence="execute"
      Value="&quot;[SystemFolder]\WindowsPowerShell\v1.0\powershell.exe&quot; -NoProfile -ExecutionPolicy Bypass -Command &quot;&amp; '[#ChocolateyInstall]' -ChocolateyDownloadUrl '[#ChocolateyNupkg]' -UseNativeUnzip&quot;" />
    <CustomAction Id="InstallChocolatey" BinaryKey="WixCA" DllEntry="WixQuietExec" Execute="deferred" Return="check" Impersonate="no" />
```

**Reply by bbmmcodegirl** [reply]
If I omit the double quotes, I get a complaint that the command must be in double quotes. So what is it? 
`&quot;[SystemFolder]\WindowsPowerShell\v1.0\powershell.exe&quot;` also has double quotes. 
In your example, it appears that you only need to quote the command itself, not its parameters. But that doesn't explain why "dir" fails which doesn't have parameters. This just absolutely has to be documented how to do.
![image](https://github.com/wixtoolset/issues/assets/6622880/b873afae-5cbf-40af-a779-41a33fefa69d)

**Reply by bevanweiss** [reply]
Rob and co have a company that provides (paid) support
https://www.firegiant.com/

**Reply by bbmmcodegirl** [reply]
Who is responsible for updating the documentation on WiX? Probably not Rob & Co?
... people would not need support if the documentation was sufficient.

**Reply by bevanweiss** [reply]
By all means, write the doco, raise a pull request... that's how open source works.

**Reply by robmen** [reply]
@bbmmcodegirl Paying for support is the most direct way companies support the WiX Toolset. Even if the documentation was perfect (and it is far from perfect), everything here (code, issues, discussions, builds, releases) still needs to be maintained. That doesn't happen magically.

I encourage you to take a moment and consider how things get done.

**Reply by bevanweiss** [reply]
@robmen Have you considered signing up to GitHub Sponsers (https://github.com/sponsors), or having a Patreon (https://www.patreon.com/pricing)?  It would provide an opportunity for individuals to financially support WiX development that I don't think is currently captured by the FireGiant support options.
